modules/InboundEmail/InboundEmail.php

```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
                     return false;
                 }
                 $email->$rel->add($email->parent_id);
-            }
+            } else {
```

Comment:
This function complexity is getting hardly, almost unreadable. Please split out into separated functions.